File: packages/python/src/massed_compute_mcp/upstream.py
"""HTTP-level helpers for the small set of upstream calls the CLI itself
makes. The MCP server's per-tool forwarder is separate (server.py)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

import httpx

TOKEN_VALIDATION_PATH = "/api/v1/account/token/validation"


@dataclass
class ValidationResult:
    status: Literal["ok", "unauthorized", "upstream_error", "network_error"]
    http_status: int = 0
    detail: str = ""


def validate_api_key(api_key: str, base_url: str, *, timeout: float = 30.0) -> ValidationResult:
    url = f"{base_url.rstrip('/')}{TOKEN_VALIDATION_PATH}"
    try:
        res = httpx.post(
            url,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            },
            json={},
            timeout=timeout,
        )
    except httpx.HTTPError as err:
        return ValidationResult(status="network_error", detail=str(err))
    if res.is_success:
        return ValidationResult(status="ok", http_status=res.status_code)
    if res.status_code in (401, 403):
        return ValidationResult(status="unauthorized", http_status=res.status_code)
    return ValidationResult(status="upstream_error", http_status=res.status_code)

File: packages/node/src/upstream.ts
/**
 * Helpers for the small set of upstream calls the CLI itself makes (the
 * token-validation request `init` and `doctor` use). The MCP server's
 * tool-call forwarder is separate (`forward.ts`) and only ever runs after
 * the CLI has handed off to the JSON-RPC dispatcher.
 */

export const enum ValidationOutcome {
  Ok = "ok",
  Unauthorized = "unauthorized",
  UpstreamError = "upstream_error",
  NetworkError = "network_error",
}

export type ValidationResult =
  | { status: ValidationOutcome.Ok; httpStatus: number }
  | { status: ValidationOutcome.Unauthorized; httpStatus: number }
  | { status: ValidationOutcome.UpstreamError; httpStatus: number }
  | { status: ValidationOutcome.NetworkError; detail: string };

const TOKEN_VALIDATION_PATH = "/api/v1/account/token/validation";

export const validateApiKey = async (
  apiKey: string,
  baseUrl: string,
): Promise<ValidationResult> => {
  const url = `${baseUrl.replace(/\/+$/, "")}${TOKEN_VALIDATION_PATH}`;
  let res: Response;
  try {
    res = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: "{}",
      // Mirror httpx's default timeout in the Python sibling — without
      // this, a stuck upstream during `init` or `doctor` would hang the
      // CLI silently.
      signal: AbortSignal.timeout(30_000),
    });
  } catch (err) {
    return {
      status: ValidationOutcome.NetworkError,
      detail: err instanceof Error ? err.message : String(err),
    };
  }
  if (res.ok) return { status: ValidationOutcome.Ok, httpStatus: res.status };
  if (res.status === 401 || res.status === 403) {
    return { status: ValidationOutcome.Unauthorized, httpStatus: res.status };
  }
  return { status: ValidationOutcome.UpstreamError, httpStatus: res.status };
};
